### PR TITLE
Added AES encryption/decryption using native crypto 

### PR DESCRIPTION
**Pull request by cbaker6**
This PR does the following:
- [x] Adds AES encryption/decryption to adapter (used #8 as a template).
- [x] Adds key rotation to move unencrypted files to encrypted, swap old keys for new ones, or remove file encryption
- [x] Use createRead/WriteStream for files instead of reading/write whole files in memory (should be a performance improvement for larger files)
- [x] Make FS and and GridFS adapters look similar. This should make it easier copy updates/changes from one adapter to the other as they are essentially doing the same thing, but to different storage mediums.
- [x] Updates travis to make it work with updated packages  

For `fileKey` can pass in `PARSE_SERVER_FILE_KEY` (fileKey) from parse-server or supply any key string you want (just be sure to keep it a secret).

Example:
```javascript
const api = new ParseServer({
  databaseURI: databaseUri || 'mongodb://localhost:27017/dev',
  cloud: process.env.PARSE_SERVER_CLOUD || __dirname + '/cloud/main.js',
  appId: process.env.PARSE_SERVER_APPLICATION_ID || 'myAppId',
  masterKey: process.env.PARSE_SERVER_MASTER_KEY || '', 
  filesAdapter: new FSFilesAdapter({fileKey: process.env.PARSE_SERVER_FILE_KEY}), //Add your file key here. Keep it secret. Note that this is needed for Postgres or anyone using FSFilesAdapter. 
  ...
});
```

Can rotate keys as well:

Periodically you may want to rotate your fileKey for security reasons. When this is the case. Initialize the file adapter with the new key and do the following in your `index.js`:

#### Files were previously unencrypted and you want to encrypt
```javascript
var FSFilesAdapter = require('@parse/fs-files-adapter');

var fsAdapter = new FSFilesAdapter({
  "filesSubDirectory": "my/files/folder", // optional
  "fileKey": "newKey" //Use the newKey
});

var api = new ParseServer({
	appId: 'my_app',
	masterKey: 'master_key',
	filesAdapter: fsAdapter
})

//This can take awhile depending on how many files and how larger they are. It will attempt to rotate the key of all files in your filesSubDirectory
const {rotated, notRotated} =  await api.filesAdapter.rotateFileKey();
console.log('Files rotated to newKey: ' + rotated);
console.log('Files that couldn't be rotated to newKey: ' + notRotated);
```

#### Files were previously encrypted with `oldKey` and you want to encrypt with `newKey`
The same process as above, but pass in your `oldKey` to `rotateFileKey()`.
```javascript
//This can take awhile depending on how many files and how larger they are. It will attempt to rotate the key of all files in your filesSubDirectory
const {rotated, notRotated} =  await api.filesAdapter.rotateFileKey({oldKey: oldKey});
console.log('Files rotated to newKey: ' + rotated);
console.log('Files that couldn't be rotated to newKey: ' + notRotated);
```

#### Only rotate a select list of files that were previously encrypted with `oldKey` and you want to encrypt with `newKey`
This is useful if for some reason there errors and some of the files werent rotated and returned in `notRotated`. The same process as above, but pass in your `oldKey` along with the array of `fileNames` to `rotateFileKey()`.
```javascript
//This can take awhile depending on how many files and how larger they are. It will attempt to rotate the key of all files in your filesSubDirectory
const {rotated, notRotated} =  await api.filesAdapter.rotateFileKey({oldKey: oldKey, fileNames: [fileName1,fileName2]});
console.log('Files rotated to newKey: ' + rotated);
console.log('Files that couldn't be rotated to newKey: ' + notRotated);
```

### CODE REVIEW

File: README.md
Comment (davimacedo) [inline README.md:86]: If you only rotate a subset of the files, how will the not rotated ones be served?

File: index.js
Comment (davimacedo) [inline index.js:10]: I didn't find it in package.json. Also, do you believe it would be possible to separate all the crypto login in another file that only would be loaded in the case the cryptography was activated?

Comment (cbaker6) [reply index.js:10]: I believe crypto is native to node like "fs". I assume this is why Flo wanted to go with it [here](https://github.com/parse-community/parse-server-fs-adapter/pull/8#issuecomment-307992864) 

> I really appreciate the effort on this PR, and that seems highly valuable. Could we use the native crypto libraries instead of node-cipher as it doesn't look heavily supported or popular and we try to limit dependencies over 3rd party libraries when possible.

For example, I never had to add it the package.json to use it and neither does the testcases

File: README.md
Comment (cbaker6) [reply README.md:86]: You are right, this is additional functionality for people who have custom setups or if the server crashes/has issues during rotation and you need to rotate the rest of the files. 

For the most part, admins should store files for respective apps in individual folders, and rotate all when needed.

File: index.js
Comment (davimacedo) [reply index.js:10]: Yes. I've just checked here. Sorry for my mistake.

File: README.md
Comment (davimacedo) [reply README.md:86]: I believe that rotating the keys like this can generate a lot of problems in the case there is a large amount of files in the storage. I'd prefer to either have this rotate functionality on a separate module/project or think in some way to have both the old and the new key serving old and new files.

**Comment by cbaker6** [comment]
@davimacedo I'll note that I made FS and GridFSBucket look almost exactly the same (testcases are slightly different, but do the samething) so they can be modified in a similar way moving forward since they do the same thing but for different file systems. The original PR was already approved, but the key rotation is [here for GridFS](https://github.com/parse-community/parse-server/pull/6768). Since you are looking at the code for this PR, it might be easier for your to comment on the changes there as well

File: README.md
Comment (cbaker6) [reply README.md:86]: I'm not sure I understand, how does encrypting keys with this module, but having the key rotation on a different module/project help? I think if you encrypt files, that same module should always have a way to decrypt them and leave unencrypted if wanted, which will probably be the main use of the rotate functionality.

Server admins can come up with their own creative ways of maintaining their key, like encrypting the fileKey with a separate key and then rotating the separate key instead of rotating the fileKey encrypting all of the files. They have to maintain two keys, but gets around the large amount of files issues (I think AWS does something similar to this). I'm sure others can come up with a ton of ways, but the ability to decrypt should always be apart of the same module IMO.

Parse-server admins determine when to rotate keys. Another way is they could create create a separate dev parse-server, rotate the keys of a copy of their files directory and then point their production server to the new folder.

Comment (davimacedo) [reply README.md:86]: I believe this rotate function shouldn't be part of the scope of the adapter and that's why I suggested either a separate project or a separate script. You are suggesting people to use the rotate function in the Parse Server initialization which can generate a lot of problems in the case there is a large amount of files. Take the example below:

```js
var FSFilesAdapter = require('@parse/fs-files-adapter');
var fsAdapter = new FSFilesAdapter({	var fsAdapter = new FSFilesAdapter({
      "filesSubDirectory": "my/files/folder" // optional	  "filesSubDirectory": "my/files/folder", // optional
    });	  "fileKey": "newKey" //Use the newKey
});
var api = new ParseServer({	var api = new ParseServer({
	appId: 'my_app',		appId: 'my_app',
	masterKey: 'master_key',		masterKey: 'master_key',
	filesAdapter: fsAdapter		filesAdapter: fsAdapter
})	})
//This can take awhile depending on how many files and how larger they are. It will attempt to rotate the key of all files in your filesSubDirectory
const {rotated, notRotated} =  await api.filesAdapter.rotateFileKey();
console.log('Files rotated to newKey: ' + rotated);
console.log('Files that couldn't be rotated to newKey: ' + notRotated);
```

- Now Imagine that I am running Parse Server in multiple containers. I will have multiple processed running the rotation at the same time.
- The rotation will consume cpu and memory in these processes which can cause poor performance while it is running.
- The files not yet rotated will not be correctly served

I still believe that the rotate function is useful though, but we should either solve all these problems (which will be hard to be done) or not suggest it to be used this way.

Comment (cbaker6) [reply README.md:86]: Hmmm... it wasn’t my intent to suggest a way how to use it, but merely show how it can be used (I believe I put a warning about performance in the Readme of FS and GridFS). I personally would not run rotate on multiple containers. Instead, one container should run it to ensure integrity and make sure multiple containers don’t  attempt to touch the same file (I also added in the readme how people can run multiple containers correctly with FS). As you mentioned, that particular container may have poor performance (not as poor as the previous version of FS because the new version uses streams), the others will run normally, but won’t be able to serve up files that haven’t rotated to the new key yet. To properly manage this, a parse server admin should modify their load balancer to move traffic away from the container rotating the key. 

But this is merely one solution, more knowledgeable admins than me will surely come up with better solutions. I don’t think this should get in the way of this PR. I simply provided a way to encrypt files, rotate keys, and decrypt files. If others wanted to create a “best practice” on how to “rotate keys” I’m all for it.

I’ll also point out, admins don’t have to rotate their keys, nor do they have to encrypt their files at all. The changes from this PR will give those admins better performance than before. For those who want to encrypt, they have to consider the trade offs mentioned here

Comment (davimacedo) [reply README.md:86]: I would probably run the rotate key operation in a single process apart of Parse Server and I would be closely monitoring that process until it had finished. That's why I believe that the rotate script is not scope of this adapter. It will lead people to use it in a wrong way. I think we should either remove the rotate function from this PR or move it to another place (maybe another project, or a gist, or even a script in readme, or even a separate script in this project - maybe an executable file?). That's what I particularly believe but @dplewis or @mtrezza can have another opinion on that.

Comment (cbaker6) [reply README.md:86]: I think you speaking to my point. People can decide how to use the functionality. I don't think it's up to use to "dictate" how people should use it, but can provide best practices. I double down on my earlier statements, I think it's irresponsible for any module to encrypt a file and not provide a way to decrypt it. Why require someone to need two modules? The crypto module doesn't do that, nor do many others. 

In the readme and testcases, I show how to encrypt, decrypt, swap keys, remove keys. People chose how and when to use them in ways that fits their use cases.

Comment (cbaker6) [reply README.md:86]: If someone wants to add improvements to rotating keys, that's definitely welcomed as well, but if they are not simple enough, seems like a different PR. As it stands, the s3 and GC file adapters encrypt files, the GridFS can encrypt now based off of the other PR I submitted, and FS can't without this PR)

Comment (davimacedo) [reply README.md:86]: The features that this and the other PR add are very welcome. I'm only questioning the rotating keys functionality, which I believe will generate further problems. We can just remove this feature from this PR and address it in a future PR. What do you think?

Comment (cbaker6) [reply README.md:86]: It's definitely possible to move out of this PR. My concern is, without functionality similar to rotateKeys, admins who encrypt their files will not be able to remove encryption of their files at rest. Of course, their files will be served decrypted when the parse-server is using the correct key, but they have no way of going back since rotatekeys also allows you to rotate from: noKey->newKey, oldKey->newKey, oldKey->noKey. They basically will only be able do: noKey->newKey.

Comment (davimacedo) [reply README.md:86]: Again, I believe that a script that does the keys rotation makes sense and has value but in my understanding it is not part of the scope of this adapter and will lead people to error. My suggestion is either to move it to somewhere else that not the adapter instance or do not add the feature in this PR.

Comment (cbaker6) [reply README.md:86]: I personally would like to wait on others to chime in as I respectfully disagree with opinion. I don’t see any evidence you have provided to not include rotate keys in this PR, but simply your opinion on how people, “may” use the rotate keys functionality when I clearly put a warning in the comments before this was ever brought up. I have recommended if you believe this can be used incorrectly (most technology can cause problems if used incorrectly) then a “best practices” document can be created. I also disagree with moving the functionality out of this module. This would imply that the ability to encrypt should be moved out this module as well. I also think it’s irresponsible to encrypt files in one module and then expect users to search for and figure out how to use a different module to remove decryption. To further validate my last point of encryption being apart of this module, I point to my reference of Flo’s comment when you asked about crypto being in package.json, https://github.com/parse-community/parse-server-fs-adapter/pull/15#discussion_r455230026

Comment (cbaker6) [reply README.md:86]: There are features in parse-server that are marked "experimental", you can mark it experimental if you are concerned with how people "may" use rotateKeys and the "possible issues" they can run into when it comes to resources if they use it incorrectly. Note they will face this same issue if they already have a running parse-server and they attempt to encrypt their already stored files for the first time. Then you can start the best practices document and even provide improvements to how/when to call rotateKeys. Doing it this way still gives them ability to decrypt their files and doesn't leave them stuck with encrypted files.

Comment (davimacedo) [reply README.md:86]: No problem. Let's wait for the others to state their opinions.

Comment (cbaker6) [reply README.md:86]: @dplewis @mtrezza can you both weigh in on this?

The difference in opinion between @davimacedo and I are expressed on the thread and I think there are pros/cons to both approaches. I don’t believe the differences are a showstopper for this PR and I think improvements can be made by others (if I see an improvement, I will make it as well).

On a related note, @davimacedo @dplewis @mtrezza I know there has been recent/past discussions about the parse-server allowing access to all files if you guess the link (which isn’t necessarily easy, but doable). I’m going off memory, but I believe the server has something like `getHostFile` (I’ll look this up when I get to my computer). I don’t know entirely how this works, but I assume it gets the file whenever someone requests the link and calls the fileAdapters underlying methods to retrieve? If this is the case (or something similar), is possible to use the fileKey to either serve an encrypted or unencrypted file based on the user access (I’m assuming some file always to be present at the link to stay inline with older versions which is why I mention the encrypted file)? I definitely haven’t investigated if this is the easiest and best way, but I wanted to start the convo

Comment (cbaker6) [reply README.md:86]: @davimacedo I know it’s been awhile since we discussed this. If I recall, you didn’t believe the key rotation should be in the adapter itself. I approached from the direction if we provide a way to encrypt, we should provide a way to decrypt (with this adapter as well as the grid adapter on parse-server). Are we able to come to some type of agreement on this? This PR along with the parse-server have been around for a few months and I’m sure my JS skills are not as good as they were when I made the PR’s.

I’m not attempting to force these in, but my assumption is people enable file encryption here will understand what they are doing and take the necessary precautions when rotating keys. People that use the S3 adapter have this benefit because AWS handles it for them https://docs.aws.amazon.com/kms/latest/developerguide/rotate-keys.html

I want to also state that file encryption isn’t enabled by default, so I’m sure many will just be using the refreshed code

Comment (davimacedo) [reply README.md:86]: Like I told you before, you did a great job here, but the only thing I don't like in this PR is when you suggest something like this:

```js
var FSFilesAdapter = require('@parse/fs-files-adapter');

var fsAdapter = new FSFilesAdapter({
  "filesSubDirectory": "my/files/folder", // optional
  "fileKey": "newKey" //Use the newKey
});

var api = new ParseServer({
	appId: 'my_app',
	masterKey: 'master_key',
	filesAdapter: fsAdapter
})

//This can take awhile depending on how many files and how larger they are. It will attempt to rotate the key of all files in your filesSubDirectory
const {rotated, notRotated} =  await api.filesAdapter.rotateFileKey();
console.log('Files rotated to newKey: ' + rotated);
console.log('Files that couldn't be rotated to newKey: ' + notRotated);
```

In my vision, the developers should never use the same process they are using for running Parse Server to also rotate the files. That's what this example is suggesting and I don't see it ending well unless the app has just few files.

I believe we can even maintain the key rotation here but it would be more clear and actually more convenient for the developers if it worked via a script that they could run something like:

```shell
parse-server-fs-adapter appId masterKey path oldKey newKey
```

If it is too much effort to be done, I believe we should at least move the rotate method from the adapter to a separate class in which you would pass appId, masterKey, path, oldKey, newKey (and not an instance of Parse Server) and make it clear in the readme that this operation is not intended to be executed in the same process of Parse Server.

The other maintainers may have a different vision though. @dplewis @mtrezza thoughts?

Comment (cbaker6) [reply README.md:86]: @davimacedo I guess my thinking was similar to your script example, with the belief that the code you mentioned (key rotation code) above would be executed on a development server (not the live server). So the development server is spun up with the `newKey`, then keys are rotated. If there are additional tests to be done on dev parse-server they can be done as well. After completion, the live server will need to be restarted with the `newKey`.

Would it help to change my examples to recommend doing key rotation on a dev server? Then mention the part about restarting the live server?

Comment (davimacedo) [reply README.md:86]: Yes. I believe it helps. Anything that makes clear to the developers that they should not run the rotate key in the same process they are running parse server.

**Comment by cbaker6** [comment]
I'm not sure what "Danger" is but it looks like something that was added to the repo after the initial PR

**Comment by davimacedo** [comment]
Yes. This Danger script is broken in all repos.

**Review by davimacedo — APPROVED** [review]
LGTM!

**Pull request merged**
